application/src/test/java/org/opentripplanner/graph_builder/module/osm/moduletests/_support/TestOsmProvider.java

```diff
@@ -0,0 +1,80 @@
+package org.opentripplanner.graph_builder.module.osm.moduletests._support;
+
+import java.time.ZoneId;
+import java.util.ArrayList;
+import java.util.List;
+import org.opentripplanner._support.time.ZoneIds;
+import org.opentripplanner.graph_builder.module.osm.OsmDatabase;
+import org.opentripplanner.osm.OsmProvider;
+import org.opentripplanner.osm.model.OsmNode;
+import org.opentripplanner.osm.model.OsmWay;
+import org.opentripplanner.osm.tagmapping.OsmTagMapper;
+import org.opentripplanner.osm.wayproperty.WayPropertySet;
+
+public class TestOsmProvider implements OsmProvider {
+
+  private final List<OsmWay> ways;
+  private final List<OsmNode> nodes;
+
+  public TestOsmProvider(List<OsmWay> ways, List<OsmNode> nodes) {
+    this.ways = ways.stream().peek(w -> w.setOsmProvider(this)).toList();
+    this.nodes = List.copyOf(nodes);
+  }
+
+  public static Builder of() {
+    return new Builder();
+  }
+
+  @Override
+  public void readOsm(OsmDatabase osmdb) {
+    ways.forEach(osmdb::addWay);
+
+    osmdb.doneSecondPhaseWays();
+
+    nodes.forEach(osmdb::addNode);
+  }
+
+  @Override
+  public OsmTagMapper getOsmTagMapper() {
+    return new OsmTagMapper();
```

Comment:
Maybe this and also WayPropertySet below could be instanced only once when the test provider is created? Graph building accesses these objects repeatedly,  so it feels wasteful to pass out a new instance every time.